File: web/src/utils/WebRtcSignalingClient.ts
/**
 * WebRTC signaling client for the web interface
 */

import { EventEmitter } from 'events';

/**
 * WebRTC signaling message types
 */
export enum SignalingMessageType {
  Register = 'register',
  Offer = 'offer',
  Answer = 'answer',
  IceCandidate = 'ice_candidate',
  Error = 'error',
}

/**
 * WebRTC signaling message
 */
export interface SignalingMessage {
  type: SignalingMessageType;
  payload: any;
}

/**
 * WebRTC signaling client events
 */
export enum SignalingClientEvent {
  Connected = 'connected',
  Disconnected = 'disconnected',
  OfferReceived = 'offer_received',
  AnswerReceived = 'answer_received',
  IceCandidateReceived = 'ice_candidate_received',
  Error = 'error',
}

/**
 * WebRTC signaling client
 */
export class WebRtcSignalingClient extends EventEmitter {
  private ws: WebSocket | null = null;
  private localPeerId: string;
  private serverUrl: string;
  private connected: boolean = false;
  private reconnectAttempts: number = 0;
  private maxReconnectAttempts: number = 5;
  private reconnectTimeout: number = 1000;

  /**
   * Create a new WebRTC signaling client
   * @param localPeerId Local peer ID
   * @param serverUrl Signaling server URL
   */
  constructor(localPeerId: string, serverUrl: string) {
    super();
    this.localPeerId = localPeerId;
    this.serverUrl = serverUrl;
  }

  /**
   * Connect to the signaling server
   */
  public async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.serverUrl);

        this.ws.onopen = () => {
          console.log('Connected to signaling server');
          this.connected = true;
          this.reconnectAttempts = 0;

          // Register with the server
          this.register();

          this.emit(SignalingClientEvent.Connected);
          resolve();
        };

        this.ws.onclose = () => {
          console.log('Disconnected from signaling server');
          this.connected = false;
          this.emit(SignalingClientEvent.Disconnected);

          // Try to reconnect
          this.reconnect();
        };

        this.ws.onerror = (error) => {
          console.error('WebSocket error:', error);
          this.emit(SignalingClientEvent.Error, error);
          reject(error);
        };

        this.ws.onmessage = (event) => {
          try {
            const message = JSON.parse(event.data) as SignalingMessage;
            this.handleMessage(message);
          } catch (error) {
            console.error('Error parsing message:', error);
          }
        };
      } catch (error) {
        console.error('Error connecting to signaling server:', error);
        reject(error);
      }
    });
  }

  /**
   * Disconnect from the signaling server
   */
  public disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
      this.connected = false;
    }
  }

  /**
   * Register with the signaling server
   */
  private register(): void {
    const message: SignalingMessage = {
      type: SignalingMessageType.Register,
      payload: {
        peer_id: this.localPeerId,
      },
    };

    this.sendMessage(message);
  }

  /**
   * Send an SDP offer to a peer
   * @param peerId Peer ID
   * @param sdp SDP offer
   */
  public sendOffer(peerId: string, sdp: string): void {
    const message: SignalingMessage = {
      type: SignalingMessageType.Offer,
      payload: {
        from: this.localPeerId,
        to: peerId,
        sdp,
      },
    };

    this.sendMessage(message);
  }

  /**
   * Send an SDP answer to a peer
   * @param peerId Peer ID
   * @param sdp SDP answer
   */
  public sendAnswer(peerId: string, sdp: string): void {
    const message: SignalingMessage = {
      type: SignalingMessageType.Answer,
      payload: {
        from: this.localPeerId,
        to: peerId,
        sdp,
      },
    };

    this.sendMessage(message);
  }

  /**
   * Send an ICE candidate to a peer
   * @param peerId Peer ID
   * @param candidate ICE candidate
   * @param sdpMid SDP mid
   * @param sdpMLineIndex SDP m-line index
   */
  public sendIceCandidate(
    peerId: string,
    candidate: string,
    sdpMid: string,
    sdpMLineIndex: number
  ): void {
    const message: SignalingMessage = {
      type: SignalingMessageType.IceCandidate,
      payload: {
        from: this.localPeerId,
        to: peerId,
        candidate,
        sdp_mid: sdpMid,
        sdp_m_line_index: sdpMLineIndex,
      },
    };

    this.sendMessage(message);
  }

  /**
   * Send a message to the signaling server
   * @param message Message to send
   */
  private sendMessage(message: SignalingMessage): void {
    if (!this.connected || !this.ws) {
      console.error('Not connected to signaling server');
      return;
    }

    try {
      this.ws.send(JSON.stringify(message));
    } catch (error) {
      console.error('Error sending message:', error);
    }
  }

  /**
   * Handle a message from the signaling server
   * @param message Message from the signaling server
   */
  private handleMessage(message: SignalingMessage): void {
    switch (message.type) {
      case SignalingMessageType.Offer:
        this.handleOffer(message.payload);
        break;
      case SignalingMessageType.Answer:
        this.handleAnswer(message.payload);
        break;
      case SignalingMessageType.IceCandidate:
        this.handleIceCandidate(message.payload);
        break;
      case SignalingMessageType.Error:
        this.handleError(message.payload);
        break;
      default:
        console.warn('Unknown message type:', message.type);
    }
  }

  /**
   * Handle an SDP offer from a peer
   * @param payload Offer payload
   */
  private handleOffer(payload: any): void {
    const { from, to, sdp } = payload;

    if (to !== this.localPeerId) {
      return;
    }

    this.emit(SignalingClientEvent.OfferReceived, from, sdp);
  }

  /**
   * Handle an SDP answer from a peer
   * @param payload Answer payload
   */
  private handleAnswer(payload: any): void {
    const { from, to, sdp } = payload;

    if (to !== this.localPeerId) {
      return;
    }

    this.emit(SignalingClientEvent.AnswerReceived, from, sdp);
  }

  /**
   * Handle an ICE candidate from a peer
   * @param payload ICE candidate payload
   */
  private handleIceCandidate(payload: any): void {
    const { from, to, candidate, sdp_mid, sdp_m_line_index } = payload;

    if (to !== this.localPeerId) {
      return;
    }

    this.emit(
      SignalingClientEvent.IceCandidateReceived,
      from,
      candidate,
      sdp_mid,
      sdp_m_line_index
    );
  }

  /**
   * Handle an error from the signaling server
   * @param payload Error payload
   */
  private handleError(payload: any): void {
    const { message } = payload;
    console.error('Signaling server error:', message);
    this.emit(SignalingClientEvent.Error, message);
  }

  /**
   * Try to reconnect to the signaling server
   */
  private reconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnect attempts reached');
      return;
    }

    this.reconnectAttempts++;
    console.log(`Reconnecting to signaling server (attempt ${this.reconnectAttempts})`);

    setTimeout(() => {
      this.connect().catch((error) => {
        console.error('Reconnect failed:', error);
      });
    }, this.reconnectTimeout * this.reconnectAttempts);
  }
}

File: darkswap-p2p/src/webrtc_signaling_client.rs
//! WebRTC signaling client wrapper
//!
//! This module provides a wrapper around the SignalingClient for WebRTC-specific functionality.

use crate::{
    error::Error,
    signaling_client::{SignalingClient, SignalingEvent},
};
use futures::{
    channel::mpsc::{self, Receiver, Sender},
    StreamExt,
};
use libp2p::PeerId;
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
};

/// WebRTC signaling client
pub struct WebRtcSignalingClient {
    /// Signaling client
    signaling_client: SignalingClient,
    /// Local peer ID
    local_peer_id: PeerId,
    /// Pending connections
    pending_connections: Arc<Mutex<HashMap<String, PendingConnection>>>,
}

/// Pending WebRTC connection
struct PendingConnection {
    /// SDP offer
    offer: Option<String>,
    /// SDP answer
    answer: Option<String>,
    /// ICE candidates
    ice_candidates: Vec<(String, String, u16)>,
    /// Connection state
    state: ConnectionState,
    /// Result sender
    result_sender: Option<Sender<Result<(), Error>>>,
}

/// WebRTC connection state
#[derive(Debug, Clone, PartialEq, Eq)]
enum ConnectionState {
    /// Waiting for offer
    WaitingForOffer,
    /// Waiting for answer
    WaitingForAnswer,
    /// Connected
    Connected,
    /// Failed
    Failed(String),
}

impl WebRtcSignalingClient {
    /// Create a new WebRTC signaling client
    pub fn new(local_peer_id: PeerId) -> Self {
        WebRtcSignalingClient {
            signaling_client: SignalingClient::new(local_peer_id.to_string()),
            local_peer_id,
            pending_connections: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Connect to the signaling server
    pub async fn connect(&self, server_url: &str) -> Result<(), Error> {
        self.signaling_client.connect(server_url).await?;
        
        // Start processing events
        self.start_event_processing().await;
        
        Ok(())
    }
    
    /// Start processing signaling events
    async fn start_event_processing(&self) {
        // Subscribe to all events
        let mut events = self.signaling_client.subscribe("*");
        
        // Clone the pending connections
        let pending_connections = self.pending_connections.clone();
        
        // Process events in a separate task
        tokio::spawn(async move {
            while let Some(event) = events.next().await {
                match event {
                    SignalingEvent::OfferReceived { from, sdp } => {
                        // Process the offer
                        let mut pending_connections = pending_connections.lock().unwrap();
                        
                        // Create or update the pending connection
                        let pending_connection = pending_connections
                            .entry(from.clone())
                            .or_insert_with(|| PendingConnection {
                                offer: None,
                                answer: None,
                                ice_candidates: Vec::new(),
                                state: ConnectionState::WaitingForOffer,
                                result_sender: None,
                            });
                        
                        // Update the offer and state
                        pending_connection.offer = Some(sdp);
                        pending_connection.state = ConnectionState::WaitingForAnswer;
                        
                        // Notify any waiting tasks
                        if let Some(sender) = &pending_connection.result_sender {
                            let _ = sender.clone().try_send(Ok(()));
                        }
                    }
                    SignalingEvent::AnswerReceived { from, sdp } => {
                        // Process the answer
                        let mut pending_connections = pending_connections.lock().unwrap();
                        
                        if let Some(pending_connection) = pending_connections.get_mut(&from) {
                            // Update the answer and state
                            pending_connection.answer = Some(sdp);
                            pending_connection.state = ConnectionState::Connected;
                            
                            // Notify any waiting tasks
                            if let Some(sender) = &pending_connection.result_sender {
                                let _ = sender.clone().try_send(Ok(()));
                            }
                        }
                    }
                    SignalingEvent::IceCandidateReceived { from, candidate, sdp_mid, sdp_m_line_index } => {
                        // Process the ICE candidate
                        let mut pending_connections = pending_connections.lock().unwrap();
                        
                        if let Some(pending_connection) = pending_connections.get_mut(&from) {
                            // Add the ICE candidate
                            pending_connection.ice_candidates.push((candidate, sdp_mid, sdp_m_line_index));
                            
                            // Notify any waiting tasks
                            if let Some(sender) = &pending_connection.result_sender {
                                let _ = sender.clone().try_send(Ok(()));
                            }
                        }
                    }
                    SignalingEvent::ErrorReceived { message } => {
                        // Process the error
                        log::error!("Signaling error: {}", message);
                        
                        // Update all pending connections
                        let mut pending_connections = pending_connections.lock().unwrap();
                        
                        for (_, pending_connection) in pending_connections.iter_mut() {
                            pending_connection.state = ConnectionState::Failed(message.clone());
                            
                            // Notify any waiting tasks
                            if let Some(sender) = &pending_connection.result_sender {
                                let _ = sender.clone().try_send(Err(Error::WebSocketError(message.clone())));
                            }
                        }
                    }
                }
            }
        });
    }
    
    /// Create a WebRTC offer
    pub async fn create_offer(&self, peer_id: &PeerId) -> Result<String, Error> {
        // Create a dummy offer for now
        // In a real implementation, this would create a WebRTC offer using the browser's RTCPeerConnection API
        let offer = format!("v=0\r\no=- 0 0 IN IP4 127.0.0.1\r\ns=-\r\nt=0 0\r\na=group:BUNDLE data\r\na=msid-semantic: WMS\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 0.0.0.0\r\na=ice-ufrag:dummy\r\na=ice-pwd:dummy\r\na=fingerprint:sha-256 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00\r\na=setup:actpass\r\na=mid:data\r\na=sctp-port:5000\r\n");
        
        // Send the offer through the signaling server
        self.signaling_client.send_offer(&peer_id.to_string(), &offer).await?;
        
        // Create or update the pending connection
        let mut pending_connections = self.pending_connections.lock().unwrap();
        
        let pending_connection = pending_connections
            .entry(peer_id.to_string())
            .or_insert_with(|| PendingConnection {
                offer: Some(offer.clone()),
                answer: None,
                ice_candidates: Vec::new(),
                state: ConnectionState::WaitingForAnswer,
                result_sender: None,
            });
        
        // Update the offer and state
        pending_connection.offer = Some(offer.clone());
        pending_connection.state = ConnectionState::WaitingForAnswer;
        
        Ok(offer)
    }
    
    /// Process a WebRTC answer
    pub async fn process_answer(&self, peer_id: &PeerId, answer: &str) -> Result<(), Error> {
        // In a real implementation, this would process a WebRTC answer using the browser's RTCPeerConnection API
        log::debug!("Processing WebRTC answer from peer {}: {}", peer_id, answer);
        
        // Update the pending connection
        let mut pending_connections = self.pending_connections.lock().unwrap();
        
        if let Some(pending_connection) = pending_connections.get_mut(&peer_id.to_string()) {
            // Update the answer and state
            pending_connection.answer = Some(answer.to_string());
            pending_connection.state = ConnectionState::Connected;
            
            // Notify any waiting tasks
            if let Some(sender) = &pending_connection.result_sender {
                let _ = sender.clone().try_send(Ok(()));
            }
        }
        
        Ok(())
    }
    
    /// Add an ICE candidate
    pub async fn add_ice_candidate(&self, peer_id: &PeerId, candidate: &str, sdp_mid: &str, sdp_m_line_index: u16) -> Result<(), Error> {
        // In a real implementation, this would add an ICE candidate using the browser's RTCPeerConnection API
        log::debug!("Adding ICE candidate for peer {}: {}", peer_id, candidate);
        
        // Send the ICE candidate through the signaling server
        self.signaling_client.send_ice_candidate(&peer_id.to_string(), candidate, sdp_mid, sdp_m_line_index).await?;
        
        // Update the pending connection
        let mut pending_connections = self.pending_connections.lock().unwrap();
        
        if let Some(pending_connection) = pending_connections.get_mut(&peer_id.to_string()) {
            // Add the ICE candidate
            pending_connection.ice_candidates.push((candidate.to_string(), sdp_mid.to_string(), sdp_m_line_index));
        }
        
        Ok(())
    }
    
    /// Wait for a connection to be established
    pub async fn wait_for_connection(&self, peer_id: &PeerId) -> Result<(), Error> {
        // Create a channel for receiving the result
        let (tx, mut rx) = mpsc::channel(1);
        
        // Update the pending connection
        {
            let mut pending_connections = self.pending_connections.lock().unwrap();
            
            let pending_connection = pending_connections
                .entry(peer_id.to_string())
                .or_insert_with(|| PendingConnection {
                    offer: None,
                    answer: None,
                    ice_candidates: Vec::new(),
                    state: ConnectionState::WaitingForOffer,
                    result_sender: None,
                });
            
            // Set the result sender
            pending_connection.result_sender = Some(tx.clone());
            
            // If the connection is already established, return immediately
            if pending_connection.state == ConnectionState::Connected {
                return Ok(());
            }
            
            // If the connection has failed, return an error
            if let ConnectionState::Failed(message) = &pending_connection.state {
                return Err(Error::WebSocketError(message.clone()));
            }
        }
        
        // Wait for the connection to be established
        match rx.next().await {
            Some(Ok(())) => Ok(()),
            Some(Err(e)) => Err(e),
            None => Err(Error::WebSocketError("Connection closed".to_string())),
        }
    }
    
    /// Get the connection state
    pub fn get_connection_state(&self, peer_id: &PeerId) -> Option<ConnectionState> {
        let pending_connections = self.pending_connections.lock().unwrap();
        
        pending_connections.get(&peer_id.to_string()).map(|c| c.state.clone())
    }
    
    /// Get the signaling client
    pub fn signaling_client(&self) -> &SignalingClient {
        &self.signaling_client
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[tokio::test]
    async fn test_webrtc_signaling_client() {
        // This test requires a running signaling server
        // It's disabled by default
        
        // Create a WebRTC signaling client
        let client = WebRtcSignalingClient::new(PeerId::random());
        
        // Connect to the signaling server
        client.connect("ws://localhost:9001/signaling").await.unwrap();
        
        // Create an offer
        let peer_id = PeerId::random();
        let offer = client.create_offer(&peer_id).await.unwrap();
        
        // Wait for the connection to be established
        // This will timeout if no answer is received
        tokio::time::timeout(std::time::Duration::from_secs(1), client.wait_for_connection(&peer_id))
            .await
            .unwrap_or(Ok(()))
            .unwrap_or(());
    }
}